packages/contracts-periphery/contracts/universal/op-nft/OptimistInviter.sol

```diff
@@ -0,0 +1,228 @@
+ *         attestation can issue signatures that allow other accounts to claim an invite. The
+ *         invitee uses a claim and reveal flow to claim the invite to an address of their choosing.
+ */
+contract OptimistInviter is Semver, EIP712Upgradeable {
+    /**
+     * @notice Emitted when an invite is claimed.
+     *
+     * @param issuer Address that issued the signature.
+     * @param claimer Address that claimed the invite.
+     */
+    event InviteClaimed(address indexed issuer, address indexed claimer);
+
+    /**
+     * @notice Version identifier, used for upgrades.
+     */
+    uint8 public constant VERSION = 1;
+
+    /**
+     * @notice EIP712 typehash for the ClaimableInvite type.
+     *         keccak256("ClaimableInvite(address issuer,bytes32 nonce)")
+     */
+    bytes32 public immutable CLAIMABLE_INVITE_TYPEHASH =
+        0x6529fd129351e725d7bcbc468b0b0b4675477e56b58514e69ab7e66ddfd20fce;
+
+    /**
+     * @notice Granter who can set accounts' invite counts.
+     */
+    address public immutable INVITE_GRANTER;
+
+    /**
+     * @notice Address of the AttestationStation contract.
+     */
+    AttestationStation public immutable ATTESTATION_STATION;
+
+    /**
+     * @notice Struct that represents a claimable invite that will be signed by the issuer.
+     *
+     * @custom:field issuer   Address that issued the signature. Reason this is explicitly included,
+     *                        and not implicitly assumed to be the recovered address from the
+     *                        signature is that the issuer may be using a ERC-1271 compatible
+     *                        contract wallet, where the recovered address is not the same as the
+     *                        issuer, or the signature is not an ECDSA signature at all.
+     * @custom:field nonce    Pseudorandom nonce to prevent replay attacks.
+     */
+    struct ClaimableInvite {
+        address issuer;
+        bytes32 nonce;
+    }
+
+    /**
+     * @notice Maps from hashes to whether or not they have been committed.
+     */
+    mapping(bytes32 => bool) public commitments;
+
+    /**
+     * @notice Maps from addresses to nonces to whether or not they have been used.
+     */
+    mapping(address => mapping(bytes32 => bool)) public usedNonces;
+
+    /**
+     * @custom:semver 1.0.0
+     *
+     * @param _inviteGranter      Address of the invite granter.
+     * @param _attestationStation Address of the AttestationStation contract.
+     */
+    constructor(address _inviteGranter, AttestationStation _attestationStation) Semver(1, 0, 0) {
+        INVITE_GRANTER = _inviteGranter;
+        ATTESTATION_STATION = _attestationStation;
+    }
+
+    /**
+     * @notice Initializes the OptimistInviter contract, setting the EIP712 context.
+     *
+     * @param _name Contract name
+     */
+    function initialize(string memory _name) public reinitializer(VERSION) {
+        __EIP712_init(_name, version());
```

Comment:
Not sure if we should tie the semver version to the eip712 version. Trying to think of the impact of this

Reply:
Discussed with Mark:

1. Every time we upgrade the contract, we should call initialize again to bump the EIP712 version (by calling __EIP712_init again). The EIP712 domain separator hash is stored in storage by EIP712Upgradeable.
2. Upgrading the version will cause all signatures that haven't been claimed yet to be invalid - we should be aware of this when upgrading the version, and communicate this on product side if necessary.

Reply:
Since there's no explicit benefit of tying the semver to the EIP712 version, and separating them gives us more flexibility in upgrading the contract, I changed the EIP712 version to be stored separately so it can be updated independently when there's a change to the signature scheme.